lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -2661,75 +2659,207 @@ struct DotOpConversionHelper {
   DotOp dot;
 };
 
-LogicalResult
-DotOpConversion::convertMMA16816(triton::DotOp op, OpAdaptor adapter,
-                                 ConversionPatternRewriter &rewriter) const {
-  Location loc = op->getLoc();
-  MLIRContext *ctx = op->getContext();
-  // D = A * B + C
-  Value A = op.a();
-  Value B = op.b();
-  Value C = op.c();
-  Value D = op.getResult();
-  bool allowTF32 = op.allowTF32();
+// This class helps to adapt the existing DotOpConversion to the latest
+// DotOpOperand layout design. It decouples the exising implementation to two
+// parts:
+// 1. loading the specific operand matrix(for $a, $b, $c) from smem
+// 2. passing the loaded value and perform the mma codegen
+struct MMA16816ConversionHelper {
+  Value A, B, C, D;
+  RankedTensorType aTensorTy, bTensorTy, dTensorTy;
+  ArrayRef<int64_t> aShape, bShape, dShape;
+  MmaEncodingAttr mmaLayout;
+  ArrayRef<unsigned int> wpt;
 
-  auto aTensorTy = A.getType().cast<RankedTensorType>();
-  auto bTensorTy = B.getType().cast<RankedTensorType>();
-  auto dTensorTy = D.getType().cast<RankedTensorType>();
+  int mmaInstrM{-1}, mmaInstrN{-1}, mmaInstrK{-1};
+  int matShapeM{-1}, matShapeN{-1}, matShapeK{-1};
+  int numRepM{-1}, numRepN{-1}, numRepK{-1};
+  Value thread, lane, warp, warpMN, warpN, warpM;
+  size_t aElemBytes{}, bElemBytes{};
 
-  auto aShape = aTensorTy.getShape();
-  auto bShape = bTensorTy.getShape();
-  auto dShape = dTensorTy.getShape();
+  DotOpConversionHelper helper;
+  triton::DotOp op;
+  DotOpAdaptor adapter;
+  ConversionPatternRewriter &rewriter;
+  TypeConverter *typeConverter;
+  Location loc;
+  MLIRContext *ctx{};
 
-  auto mmaLayout = dTensorTy.getEncoding().cast<MmaEncodingAttr>();
+  using ValueTable = std::map<std::pair<unsigned, unsigned>, Value>;
+
+  MMA16816ConversionHelper(triton::DotOp op, Value thread, DotOpAdaptor adapter,
+                           ConversionPatternRewriter &rewriter,
+                           TypeConverter *typeConverter, Location loc)
+      : helper(op), op(op), adapter(adapter), rewriter(rewriter),
+        typeConverter(typeConverter), loc(loc), ctx(op.getContext()),
+        thread(thread) {
+    A = op.a();
+    B = op.b();
+    C = op.c();
+    D = op.c();
```

Comment:
Although `C.shape == D.shape`, from semantic strictness perspective, should we make `D =op.d();`?

Reply:
Yeah, my mistake, it should be `D=op.getResult()` here.